### ISSUE TITLE
Adding types is very complicated

### ISSUE THREAD

**Issue by Leseratte10** [Vector35/binaryninja-api#1148]
I am using the "types" view in Binary NInja to define custom data types (structs, enums etc.) used in the program I analyze. 

However, the menus to create new types are very unintuitive and don't work properly. Take this type for example: 

`typedef void (*qr2_keylistcallback_t)(qr2_key_type keytype, qr2_keybuffer_t keybuffer, void *userdata);`

(qr2_key_type and qr2_keybuffer_t are already defined)

How do I enter this properly into the "create new type" window? I tried "void" as type name and the rest as "Type (C syntax)" but it says "invalid input". 

What I do then, is, I create a dummy type (say, `typedef abc int32_t`, and then right-click and select "edit type", and can then insert the whole type. 

Same for structs. Why can't there be one text box where I insert a struct in C form like this?

```struct sockaddr {
  u8_t        sa_len;
  sa_family_t sa_family;
  char        sa_data[14];
};
```
why do I need to create the struct empty and then add all the members one by one?

What I would like, is a new menu point besides "create new type", "create new structure" and "create new union" which would just be named "Create from C code" where I then just insert the complete C definition, so I don't have to edit existing types just so my custom ones work ...

**Comment by bpotchik** [comment]
What you are asking for already exists. Try both 'i' and 's' hotkeys to insert a new type. There are also context menu options.

**Comment by Leseratte10** [comment]
I know the context menu options and both hotkeys. 

"s" for structure only lets me enter name and size. No "contents". I have to manually add every single structure variable. 

And "i" works for simple types like `typedef my_custom_type int` but not for complex ones like the one I posted above.

**Comment by bpotchik** [comment]
It's supposed to be C syntax, however try leaving out the struct name from the definition. Put the name in the name field only.

**Comment by Leseratte10** [comment]
OK. Taking this struct as an example: 

```
struct test
{
	char gamename[64];
	qr2_serverkeycallback_t server_key_callback;
	int qport;
};
```

The menu point "Create struct" is useless, because I can't add the contents there. And the menu point "Create type" always gives me syntax errors. 

No matter if I enter "struct test" or "test" as name, and no matter if I write the whole definition in "Type (C Syntax)", or leave out the "struct test", or leave out that and the { } brackets, I always get a syntax error.

**Comment by bpotchik** [comment]
For struct, did you try

```struct
{
	char gamename[64];
	uint32_t server_key_callback;
	int qport;
};```

**Comment by Leseratte10** [comment]
Yes. No matter if I use "struct test" or "test" as name, with exactly those contents (except for the three backticks at the end which I suppose are there by mistake), I get "input:1: error: syntax error"

**Comment by bpotchik** [comment]
Formatting error, you will need the struct keyword without the typename.

**Comment by Leseratte10** [comment]
Thanks, that worked. But why is this feature so "hidden" / implemented so weirdly? First, I'd expect the menu point "Create new Structure" to work if I'd like to create a structure (why use "Type" instead?), second, why that weird syntax without struct name, and not just one large text box with the whole syntax?

And do you also have an idea about how I have to enter the typedef so it works?

**Comment by bpotchik** [comment]
Yeah, I think we have a round tripping issue now with types which is planned to be fixed. I made the same mistake when I first tried to create a struct and kind of got used to the broken syntax :|.

As for typedef...I just tried this and it seemed to work.

```
void (*)(uint32_t keytype, void* keybuffer, void *userdata);
```
You put the typename in the name field again.

**Comment by Leseratte10** [comment]
Thanks, that works as well. 

But I'd like to leave this issue open because even though my individual problem is now "fixed" (= has a workaround), it is still way more complicated than it would need to be. Or is there already another issue that describes the same issue?

**Comment by bpotchik** [comment]
Agreed! I think these issues will be resolved as part of the 1.2 release.

**Comment by psifertex** [comment]
Would the majority of your issues be solved if the name was used from your original pasted definition? Or are you proposing other changes too?

**Comment by psifertex** [comment]
Right -- that's what I mean, I'm wondering if that's the only change required or if there are others that could help.

**Comment by Leseratte10** [comment]
If using the name would work for 
- typedef (as in `typedef void (*qr2_keylistcallback_t)(qr2_key_type keytype, qr2_keybuffer_t keybuffer, void *userdata);`), 
- structs (as in 
```
struct test
{
	char gamename[64];
	qr2_serverkeycallback_t server_key_callback;
	int qport;
};
``` 
and
- enums (as in 
```
typedef enum 
{
	a,	
	b,	
	c,	
	d
} test;
```

that should be it. Basically, I'd like to be able to just copy one definition from a C header file into a text box in Binary Ninja and add it, and not manually figure out which part I now have to leave out (which for enums would be "enum" instead of the actual enum name, making that even more confusing). 

In addition to that, being able to add comments to the lines view would be cool (at the moment, C comments are removed when adding a type). But that is already reported in #1124.

**Comment by plafosse** [comment]
As of build v1.1.1453 you can edit existing structures and enumerations via C in the types view just click y on the first line of the struct. I have one more fix I'd like to get in which should improve initial structure creation also
